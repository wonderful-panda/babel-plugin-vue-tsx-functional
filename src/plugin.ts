import * as babel from "@babel/core";
import * as t from "@babel/types";

const FUNCNAME = "__VueFC__";

/**
 * Build object literal represents functional component.
 *
 * Input: `
 *   ctx => (<div>{ctx.props.text}</div>)
 * `
 * Output: `
 *   {
 *     functional: true,
 *     render(h, ctx) {
 *       return <div>{ctx.props.text}</div>;
 *     }
 *   }
 * `
 */
function buildFunctionalComponentObject(renderFn: t.ArrowFunctionExpression) {
  let body: t.BlockStatement;
  if (t.isBlockStatement(renderFn.body)) {
    body = renderFn.body;
  } else {
    body = t.blockStatement([t.returnStatement(renderFn.body)]);
  }
  const ret = t.objectExpression([
    t.objectProperty(t.identifier("functional"), t.booleanLiteral(true)),
    t.objectMethod(
      "method",
      t.identifier("render"),
      [t.identifier("h"), ...renderFn.params],
      body
    )
  ]);
  return ret;
}

/**
 * Get function name or method name from CallExpression.
 */
function getCalleeName(node: t.CallExpression): string | undefined {
  const callee = node.callee;
  if (t.isIdentifier(callee)) {
    // xxx(...)
    // ^^^
    return callee.name;
  }
  if (t.isMemberExpression(callee)) {
    const property = callee.property;
    if (t.isIdentifier(property)) {
      // xxx.yyy(...)
      //     ^^^
      return property.name;
    }
    if (t.isStringLiteral(property)) {
      // xxx["yyy"](...)
      //      ^^^
      return property.value;
    }
  }
  return undefined;
}

/**
 * Check specified path is assigned to variable or property, or exported
 *
 * expected patterns:
 *  [export] [const|let|var] VariableName = <node>
 *  export default <node>
 *  obj.PropertyName = <node>
 *  obj["PropertyName"] = <node>
 */
function isAssignedOrExported(path: babel.NodePath<t.CallExpression>): boolean {
  const parent = path.parentPath.node;

  if (t.isAssignmentExpression(parent) && parent.right === path.node) {
    return true;
  } else if (t.isVariableDeclarator(parent)) {
    return true;
  } else if (t.isExportDefaultDeclaration(parent)) {
    return true;
  } else {
    return false;
  }
}

/**
 * Replace __VueFC__ call expression by object literal which represents
 * functional component.
 */
function processCallExpression(path: babel.NodePath<t.CallExpression>) {
  const { node } = path;
  const funcname = getCalleeName(node);
  if (funcname !== FUNCNAME) {
    return;
  }
  if (node.arguments.length !== 1) {
    throw path.buildCodeFrameError(`${FUNCNAME} must have exactly 1 argument`);
  }
  const arg = node.arguments[0];
  if (!t.isArrowFunctionExpression(arg)) {
    throw path.buildCodeFrameError(
      `argument of ${FUNCNAME} must be arrow function`
    );
  }

  if (!isAssignedOrExported(path)) {
    throw path.buildCodeFrameError(
      `result of ${FUNCNAME} must be assigned or exported`
    );
  }
  const functionalComponent = buildFunctionalComponentObject(arg);
  path.replaceWith(functionalComponent);
}

export = function(): babel.PluginObj {
  const plugin: babel.PluginObj = {
    name: "vue-tsx-functional",
    visitor: {
      Program(path) {
        path.traverse({
          CallExpression(path) {
            processCallExpression(path);
          }
        });
      }
    }
  };
  return plugin;
};
