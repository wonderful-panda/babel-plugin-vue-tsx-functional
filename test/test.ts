import * as babel from "@babel/core";
import jsx from "@babel/plugin-syntax-jsx";
import plugin, { Options } from "../lib/plugin";

function transform(code: string, opts?: Options): string | undefined {
  const transformOpts: babel.TransformOptions = {
    plugins: [[plugin, opts], jsx]
  };
  const ret = babel.transform(code, transformOpts);
  return ret ? ret.code || undefined : undefined;
}

describe("Forms of render functions: ", () => {
  it("() => <div />", () => {
    const code = transform(
      `const MyComponent = __VueFC__(() => <div class='test' />)`
    );
    expect(code).toMatchSnapshot();
  });
  it("ctx => <div>...</div>", () => {
    const code = transform(
      `const MyComponent = __VueFC__(ctx => <div class='test'>{ctx.props.foo}</div>)`
    );
    expect(code).toMatchSnapshot();
  });
  it("({ ...complex... }) => <div>...</div>", () => {
    const code = transform(
      `const MyComponent = __VueFC__(({ props: { foo }, data }) => <div class='test'>{foo}</div>)`
    );
    expect(code).toMatchSnapshot();
  });

  it("() => { return <div /> }", () => {
    const code = transform(
      `const MyComponent = __VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });
});

describe("Forms of marker function: ", () => {
  it("xxx.__VueFC__(...)", () => {
    const code = transform(
      `const MyComponent = babel_plugin_vue_tsx_functional_1.__VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });
  it("xxx.yyy.__VueFC__(...)", () => {
    const code = transform(
      `const MyComponent = xxx.yyy.__VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });
  it("xxx['__VueFC__'](...)", () => {
    const code = transform(
      `const MyComponent = xxx["__VueFC__"](() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });
  it("Changing marker name", () => {
    const code = transform(
      `const MyComponent1 = _f(() => <div />);
      const MyComponent2 = xx._f(() => <div />);
      const MyComponent3 = __VueFC__(() => <div />);`,
      { funcName: "_f" }
    );
    expect(code).toMatchSnapshot();
  });
});

describe("Export or assignment: ", () => {
  it("export const MyComponent = ...", () => {
    const code = transform(
      `export const MyComponent = __VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });

  it("export default ...", () => {
    const code = transform(
      `export default __VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });

  it("exports.MyCompoent = ...", () => {
    const code = transform(
      `exports.MyComponent = __VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });

  it("let MyCompoent; MyComponent = ...", () => {
    const code = transform(
      `let MyComponent;
      MyComponent = __VueFC__(() => {
        return <div class='test' />;
      })`
    );
    expect(code).toMatchSnapshot();
  });
});
