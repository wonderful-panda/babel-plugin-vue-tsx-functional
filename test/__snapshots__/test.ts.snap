// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export or assignment:  export const MyComponent = ... 1`] = `
"export const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Export or assignment:  export default ... 1`] = `
"export default {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Export or assignment:  exports.MyCompoent = ... 1`] = `
"exports.MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Export or assignment:  let MyCompoent; MyComponent = ... 1`] = `
"let MyComponent;
MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of marker function:  Changing marker name 1`] = `
"const MyComponent1 = {
  functional: true,

  render(h) {
    return <div />;
  }

};
const MyComponent2 = {
  functional: true,

  render(h) {
    return <div />;
  }

};

const MyComponent3 = __VueFC__(() => <div />);"
`;

exports[`Forms of marker function:  xxx.__VueFC__(...) 1`] = `
"const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of marker function:  xxx.yyy.__VueFC__(...) 1`] = `
"const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of marker function:  xxx['__VueFC__'](...) 1`] = `
"const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of render functions:  () => <div /> 1`] = `
"const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of render functions:  () => { return <div /> } 1`] = `
"const MyComponent = {
  functional: true,

  render(h) {
    return <div class='test' />;
  }

};"
`;

exports[`Forms of render functions:  ({ ...complex... }) => <div>...</div> 1`] = `
"const MyComponent = {
  functional: true,

  render(h, {
    props: {
      foo
    },
    data
  }) {
    return <div class='test'>{foo}</div>;
  }

};"
`;

exports[`Forms of render functions:  ctx => <div>...</div> 1`] = `
"const MyComponent = {
  functional: true,

  render(h, ctx) {
    return <div class='test'>{ctx.props.foo}</div>;
  }

};"
`;
